boot/bootutil/include/bootutil/fault_injection_hardening.h

```diff
@@ -0,0 +1,339 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2020 Arm Limited
+ */
+
+#ifndef __FAULT_INJECTION_HARDENING_H__
+#define __FAULT_INJECTION_HARDENING_H__
+
+/* Fault injection mitigation library.
+ *
+ * Has support for different measures, which can either be enabled/disabled
+ * separately or by defining one of the MCUBOOT_FIH_PROFILEs.
+ *
+ * NOTE: These constructs against fault injection attacks are not guaranteed to
+ *       be secure for all compilers, but execution is going to be correct and
+ *       including them will certainly help to harden the code.
+ *
+ * DOUBLE_VARS makes critical variables into a tuple (x, x ^ msk). Then the
+ * correctness of x can be checked by XORing the two tuple values together. This
+ * also means that comparisons between fih_ints can be verified by doing
+ * x == y && x_msk == y_msk.
+ *
+ * GLOBAL_FAIL makes all while(1) failure loops redirect to a global failure
+ * loop. This loop has mitigations against loop escapes / unlooping. This also
+ * means that any unlooping won't immediately continue executing the function
+ * that was executing before the failure.
+ *
+ * CFI (Control Flow Integrity) creates a global counter that is incremented
+ * before every FIH_CALL of vulnerable functions. On the function return the
+ * counter is decremented, and after the return it is verified that the counter
+ * has the same value as before this process. This can be used to verify that
+ * the function has actually been called.
+ *
+ * DELAY causes random delays. This makes it hard to cause faults precisely.
+ * DELAY requires an RNG. An mbedtls integration is provided in
+ * fault_injection_hardening_delay_mbedtls.h, but any RNG that has an entropy
+ * source can be used by implementing the fih_delay_random_uchar function.
+ *
+ * The basic call pattern is:
+ *
+ * fih_int fih_rc = FIH_FAILURE;
+ * FIH_CALL(vulnerable_function, fih_rc, arg1, arg2);
+ * if (fih_not_eq(fih_rc, FIH_SUCCESS)) {
+ *     FIH_PANIC;
+ * }
+ *
+ * Note that any function called by FIH_CALL must only return using FIH_RETURN,
+ * as otherwise the CFI counter will not be decremented and the CFI check will
+ * fail causing a panic.
+ */
+
+#include "mcuboot_config/mcuboot_config.h"
+
+#if defined(MCUBOOT_FIH_PROFILE_MAX)
+
+#define FIH_ENABLE_DELAY         /* Requires an entropy source */
+#define FIH_ENABLE_DOUBLE_VARS
+#define FIH_ENABLE_GLOBAL_FAIL
+#define FIH_ENABLE_CFI
+
+#elif defined(MCUBOOT_FIH_PROFILE_MEDIUM)
```

Comment:
IMO Using both `MODERATE` and `MEDIUM` could be very confusing. I think something like `OFF / LOW / MEDIUM / HIGH` would be better.

Reply:
Agreed

Reply:
Fixed.